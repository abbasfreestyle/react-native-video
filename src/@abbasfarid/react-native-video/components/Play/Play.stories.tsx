import React from 'react';
import { Text } from 'react-native';

import { text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react-native';

import { Container } from '../../../../../storybook/helpers';

storiesOf('Components', module).add('Play Button', () => {
  return (
    <Container>
      <Text>{text('Text', 'Hello')}</Text>
    </Container>
  );
});
