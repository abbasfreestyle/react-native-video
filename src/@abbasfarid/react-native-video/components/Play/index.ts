export { Play, Props as PlayProps } from './Play';
