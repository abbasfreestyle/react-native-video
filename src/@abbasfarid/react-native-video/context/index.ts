export { useVideo, VideoProvider } from './video';
