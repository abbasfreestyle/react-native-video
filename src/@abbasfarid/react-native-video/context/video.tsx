import React, { createContext, ReactNode, useContext, useState } from 'react';

type IVideoProvider = {
  display: boolean;
  play: boolean;
  toggleDisplay: () => void;
  togglePlay: () => void;
};

const noop = () => {
  // do nothing;
};

/* istanbul ignore next */
const initialState: IVideoProvider = {
  display: false,
  play: false,
  toggleDisplay: noop,
  togglePlay: noop,
};

const VideoContext = createContext<IVideoProvider>(initialState);

export const VideoProvider = ({ children }: { children: ReactNode }) => {
  const value = useVideoProvider();
  return (
    <VideoContext.Provider value={value}>{children}</VideoContext.Provider>
  );
};

const useVideoProvider = () => {
  const [play, setPlay] = useState(false);
  const [display, setDisplay] = useState(false);

  const togglePlay = () => {
    setPlay((prev) => !prev);
  };

  const toggleDisplay = () => {
    setDisplay((prev) => !prev);
  };

  return {
    display,
    play,
    togglePlay,
    toggleDisplay,
  };
};

export const useVideo = () => useContext(VideoContext);
