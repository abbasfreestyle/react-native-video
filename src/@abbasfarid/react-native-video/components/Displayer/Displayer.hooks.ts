import { useEffect } from 'react';

import {
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from 'react-native-reanimated';

import { useVideo } from '../../context';
import { useDebounce } from '../../hooks';

export const useOpacity = (display: boolean) => {
  const opacity = useSharedValue(0);
  const style = useAnimatedStyle(() => ({
    opacity: withTiming(opacity.value, { duration: 200 }),
  }));

  useEffect(() => {
    opacity.value = display ? 1 : 0;
  }, [display, opacity]);

  return { style };
};

export const useDisplayer = () => {
  const { display, play, toggleDisplay } = useVideo();
  const { clearDebounce, debounce } = useDebounce();

  useEffect(() => {
    if (!display || !play) clearDebounce();

    if (display && play) debounce(toggleDisplay, 2000);
  }, [display, play, clearDebounce, toggleDisplay, debounce]);

  return { display, toggleDisplay };
};
