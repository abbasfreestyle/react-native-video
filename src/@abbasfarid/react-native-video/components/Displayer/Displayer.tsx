import React, { ReactNode, useEffect } from 'react';
import { Pressable, StyleSheet } from 'react-native';

import Animated, {
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from 'react-native-reanimated';

import { useVideo } from '../../context';
import { useDebounce } from '../../hooks';
import { Skip } from '../Skip';

/* 
layers:

[title, play, slider]
*[rewind, fast-fwd]
  on double-tap (maybe triple >> max taps) -> time +- X seconds
displayer (children)
  beginning/end -> visible: true
video
*/

type Props = {
  children?: ReactNode;
  backgroundColor?: string;
};

const useOpacity = (display: boolean) => {
  const opacity = useSharedValue(0);
  const style = useAnimatedStyle(() => ({
    opacity: withTiming(opacity.value, { duration: 200 }),
  }));

  useEffect(() => {
    opacity.value = display ? 1 : 0;
  }, [display, opacity]);

  return { style };
};

const BG = ({
  backgroundColor,
  display,
}: {
  display: boolean;
  backgroundColor: string;
}) => {
  const { style } = useOpacity(display);
  return (
    <Animated.View
      style={[style, styles.background, { backgroundColor }]}
      testID="animated-displayer-bg"
    />
  );
};

export const Displayer = ({
  backgroundColor = 'rgba(0,0,0,0.25)',
  children,
}: Props) => {
  const { display, play, toggleDisplay } = useVideo();
  const { clearDebounce, debounce } = useDebounce();

  const { style } = useOpacity(display);

  useEffect(() => {
    if (!display || !play) clearDebounce();

    if (display && play) debounce(toggleDisplay, 2000);
  }, [display, play, clearDebounce, toggleDisplay, debounce]);

  return (
    <>
      <BG backgroundColor={backgroundColor} display={display} />
      <Pressable
        accessibilityRole="button"
        onPress={toggleDisplay}
        style={styles.container}
      >
        <Skip direction="rewind" />
        <Animated.View style={style} testID="animated-displayer">
          {children}
        </Animated.View>
        <Skip direction="fast-forward" />
      </Pressable>
    </>
  );
};

const styles = StyleSheet.create({
  background: {
    ...StyleSheet.absoluteFillObject,
  },
  container: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
  },
});
