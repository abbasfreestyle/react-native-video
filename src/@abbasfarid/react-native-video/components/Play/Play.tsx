import React, { ComponentProps, useEffect } from 'react';
import { Pressable, PressableProps, StyleSheet, View } from 'react-native';

import Animated, {
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from 'react-native-reanimated';

import { useVideo } from '../../context';
import { Icon } from '../Icon';

export type Props = {
  play: boolean;
  onPress?: PressableProps['onPress'];
};

const iconProps: Omit<ComponentProps<typeof Icon>, 'type'> = {
  size: 50,
};

export const Play = () => {
  const { display, play, togglePlay } = useVideo();
  const transform = useSharedValue(1);

  const animatedStyle = useAnimatedStyle(() => ({
    transform: [
      {
        scale: withTiming(transform.value, { duration: 100 }, (finished) => {
          /* istanbul ignore next */
          if (finished) transform.value = 1;
        }),
      },
    ],
  }));

  useEffect(() => {
    transform.value = 1.15;
  }, [play, transform]);

  return (
    <View style={styles.container}>
      <Pressable
        accessibilityRole="button"
        disabled={!display}
        onPress={togglePlay}
        style={styles.button}
      >
        <Animated.View style={animatedStyle} testID="rn-video-play-button">
          <Icon
            testID="play-icon"
            type={play ? 'Pause' : 'Play'}
            {...iconProps}
          />
        </Animated.View>
      </Pressable>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  button: {
    height: 60,
    width: 60,
    alignItems: 'center',
    justifyContent: 'center',
  },
});
