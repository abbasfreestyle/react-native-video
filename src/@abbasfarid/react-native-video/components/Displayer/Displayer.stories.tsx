import React from 'react';
import { StyleSheet, View } from 'react-native';

import { storiesOf } from '@storybook/react-native';

import { Play, VideoProvider } from '../..';
import { Displayer } from './Displayer';

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
});

const ToggleDisplayer = () => {
  return (
    <VideoProvider>
      <View style={styles.container}>
        <Displayer>
          <Play />
        </Displayer>
      </View>
    </VideoProvider>
  );
};

storiesOf('Components', module).add('Displayer', () => <ToggleDisplayer />);
