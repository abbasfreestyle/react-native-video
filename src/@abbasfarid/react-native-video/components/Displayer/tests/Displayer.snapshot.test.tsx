import React from 'react';

import { render } from '@testing-library/react-native';

import { MockVideoProvider } from '../../../context';
import { Play } from '../../Play';
import { Displayer } from '../Displayer';

const renderTestComponent = () =>
  render(renderComponent(), { wrapper: MockVideoProvider });

const renderComponent = () => (
  <Displayer>
    <Play />
  </Displayer>
);

test('displayer should match the snapshot', () => {
  const { toJSON } = renderTestComponent();
  expect(toJSON()).toMatchSnapshot();
});
