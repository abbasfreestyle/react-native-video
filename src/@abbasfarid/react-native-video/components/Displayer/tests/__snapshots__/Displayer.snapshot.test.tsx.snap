// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displayer should match the snapshot 1`] = `
Array [
  <View
    animatedStyle={
      Object {
        "value": Object {
          "opacity": 0,
        },
      }
    }
    collapsable={false}
    style={
      Object {
        "backgroundColor": "rgba(0,0,0,0.25)",
        "bottom": 0,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    testID="animated-displayer-bg"
  />,
  <View
    accessibilityRole="button"
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <View
          animatedStyle={
            Object {
              "value": Object {
                "opacity": 0,
              },
            }
          }
          collapsable={false}
          style={
            Object {
              "opacity": 0,
              "position": "absolute",
            }
          }
        >
          <SvgMock
            color="white"
            height={25}
            width={25}
          />
        </View>
        <Text
          animatedStyle={
            Object {
              "value": Object {
                "opacity": 0,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                ],
              },
            }
          }
          collapsable={false}
          style={
            Object {
              "color": "white",
              "fontSize": 12,
              "opacity": 0,
              "position": "absolute",
              "textAlign": "center",
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          0 sec
        </Text>
        <View
          animatedStyle={
            Object {
              "value": Object {
                "opacity": 0,
                "transform": Array [
                  Object {
                    "scale": 0.1,
                  },
                ],
              },
            }
          }
          collapsable={false}
          style={
            Object {
              "backgroundColor": "white",
              "borderRadius": 50,
              "height": 100,
              "opacity": 0,
              "position": "absolute",
              "transform": Array [
                Object {
                  "scale": 0.1,
                },
              ],
              "width": 100,
            }
          }
        />
      </View>
    </View>
    <View
      animatedStyle={
        Object {
          "value": Object {
            "opacity": 0,
          },
        }
      }
      collapsable={false}
      style={
        Object {
          "opacity": 0,
        }
      }
      testID="animated-displayer"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "disabled": true,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "height": 60,
              "justifyContent": "center",
              "width": 60,
            }
          }
        >
          <View
            animatedStyle={
              Object {
                "value": Object {
                  "transform": Array [
                    Object {
                      "scale": 1,
                    },
                  ],
                },
              }
            }
            collapsable={false}
            style={
              Object {
                "transform": Array [
                  Object {
                    "scale": 1,
                  },
                ],
              }
            }
            testID="rn-video-play-button"
          >
            <SvgMock
              height={50}
              testID="play-icon"
              width={50}
            />
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <View
          animatedStyle={
            Object {
              "value": Object {
                "opacity": 0,
              },
            }
          }
          collapsable={false}
          style={
            Object {
              "opacity": 0,
              "position": "absolute",
            }
          }
        >
          <SvgMock
            color="white"
            height={25}
            width={25}
          />
        </View>
        <Text
          animatedStyle={
            Object {
              "value": Object {
                "opacity": 0,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                ],
              },
            }
          }
          collapsable={false}
          style={
            Object {
              "color": "white",
              "fontSize": 12,
              "opacity": 0,
              "position": "absolute",
              "textAlign": "center",
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          0 sec
        </Text>
        <View
          animatedStyle={
            Object {
              "value": Object {
                "opacity": 0,
                "transform": Array [
                  Object {
                    "scale": 0.1,
                  },
                ],
              },
            }
          }
          collapsable={false}
          style={
            Object {
              "backgroundColor": "white",
              "borderRadius": 50,
              "height": 100,
              "opacity": 0,
              "position": "absolute",
              "transform": Array [
                Object {
                  "scale": 0.1,
                },
              ],
              "width": 100,
            }
          }
        />
      </View>
    </View>
  </View>,
]
`;
